neuralop/layers/codano.py

```diff
@@ -0,0 +1,607 @@
+import torch.nn.functional as F
+from .fno_block import FNOBlocks
+from .spectral_convolution import SpectralConv1d, SpectralConv2d, SpectralConv3d
+
+
+# Codomain Attention Blocks
+class CODABlocks(nn.Module):
+    """ Class for the Co-domain Attention Blocks (CODABlocks)
+
+    Args:
+        n_modes (list): Number of modes for each dimension  used in K,Q,V operator.
+        n_head (int): Number of heads for the attention mechanism.
+        token_codimension (int): Co-dimension of each variable or number of channels associated with each variables.
+        output_scaling_factor (float): Scaling factor for the output.
+        incremental_n_modes (list): Incremental number of modes for each dimension (for incremental training).
+        head_codimension (int): Co-dimension of each of output token for each head.
+
+        # config for FNO_blocks used as K,Q,V operator
+        use_channel_mlp (bool): whether to use mlp layers to parameterize skip connections, by default False
+        non_linearity (callable): Non-linearity function to be used.
+        preactivation (bool): whether to use preactivation, by default False.
+        fno_skip (str): Type of skip connection to be used, by default 'linear'.
+        mlp_skip (str):  module to use for ChannelMLP skip connections, by default "soft-gating".
+        channel_mlp_expansion (float): expansion parameter for self.channel_mlp, by default 0.5.
+        separable (bool): whether to use separable convolutions, by default False.
+        factorization (str): Type of factorization to be used, by default 'tucker'.
+        rank (float): Rank of the factorization, by default 1.0.
+        SpectralConvolution (callable): Spectral convolution module to be used.
+        joint_factorization (bool): whether to factorize all spectralConv weights as one tensor, by default False
+
+        # Normalization
+        Normalizer (callable): Normalization module to be used.
+
+        codimension_size (int): Size of the codimension for whole function, only used for permutation_eq = False.
+        per_channel_attention (bool): whether to use per channel attention, by default True (overwrite token_codimension to 1).
+        permutation_eq (bool): whether to use permutation equivariant mixer layer  after attention mechanism.
+        temperature (float): Temperature parameter for attention mechanism.
+        kqv_non_linear (bool): whether to use non-linear activation for K,Q,V operator.
+    """
+
+    def __init__(
+        self,
+        n_modes,
+        n_head=1,
+        token_codimension=1,
+        output_scaling_factor=None,
+        incremental_n_modes=None,
+        head_codimension=None,
+        use_channel_mlp=False,
+        non_linearity=F.gelu,
+        preactivation=False,
+        fno_skip='linear',
+        channel_mlp_skip='soft-gating',
+        channel_mlp_expansion=1.0,
+        separable=False,
+        factorization='tucker',
+        rank=1.0,
+        SpectralConvolution=None,
```

Comment:
ideally change this arg to something other than the class name

Reply:
ok.